File: ios/objNative/JSAMQPManager.m
//
//  JSAMQPManager.m
//  Supermap
//
//  Created by 王子豪 on 2017/5/12.
//  Copyright © 2017年 Facebook. All rights reserved.
//

#import "JSAMQPManager.h"

#import "SuperMap/AMQPManager.h"
#import "JSObjManager.h"
@implementation JSAMQPManager
RCT_EXPORT_MODULE();
RCT_REMAP_METHOD(createObj,createObjWithresolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        AMQPManager* manager = [[AMQPManager alloc]init];
        NSInteger nsKey = (NSInteger)manager;
        [JSObjManager addObj:manager];
        resolve(@{@"_AMQPManagerId":@(nsKey).stringValue});
    } @catch (NSException *exception) {
        reject(@"AMQPManager",@"create Obj expection",nil);
    }
}

RCT_REMAP_METHOD(newReceiver,newReceiverById:(NSString*)managerId queueName:(NSString*)name resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        AMQPManager* manager = [JSObjManager getObjWithKey:managerId];
        AMQPReceiver* receiver = [manager newReceiver:name];
        NSInteger nsKey = (NSInteger)receiver;
        [JSObjManager addObj:receiver];
        resolve(@{@"AMQPReceiverId":@(nsKey).stringValue});
    } @catch (NSException *exception) {
        reject(@"AMQPManager",@"new Receiver expection",nil);
    }
}

RCT_REMAP_METHOD(newSender,newSenderById:(NSString*)managerId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        AMQPManager* manager = [JSObjManager getObjWithKey:managerId];
        AMQPSender* sender = [manager newSender];
        NSInteger nsKey = (NSInteger)sender;
        [JSObjManager addObj:sender];
        resolve(@{@"AMQPSenderId":@(nsKey).stringValue});
    } @catch (NSException *exception) {
        reject(@"AMQPManager",@"new Sender expection",nil);
    }
}

RCT_REMAP_METHOD(connection,connectionById:(NSString*)managerId paramDic:(NSDictionary*)paramDic resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        AMQPManager* manager = [JSObjManager getObjWithKey:managerId];
        NSString* ip = [paramDic objectForKey:@"IP"];
        int port = ((NSNumber*)[paramDic objectForKey:@"Port"]).intValue;
        NSString* hostName = [paramDic objectForKey:@"HostName"];
        NSString* userName = [paramDic objectForKey:@"UserName"];
        NSString* passWord = [paramDic objectForKey:@"PassWord"];
        NSString* clientId = [paramDic objectForKey:@"ClientId"];
        BOOL isConnection = [manager connection:ip port:port hostname:hostName usrname:userName password:passWord clientId:clientId];
        NSNumber* nsConnection = [NSNumber numberWithBool:isConnection];
        resolve(@{@"isConnection":nsConnection});
    } @catch (NSException *exception) {
        reject(@"AMQPManager",@"connection expection",nil);
    }
}

RCT_REMAP_METHOD(disconnection,disconnectionById:(NSString*)managerId resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject){
    @try {
        AMQPManager* manager = [JSObjManager getObjWithKey:managerId];
        [manager disconnection];
        resolve(@"disconnection");
    } @catch (NSException *exception) {
        reject(@"AMQPManager",@"disconnection expection",nil);
    }
}
@end

File: android/src/main/java/com/supermap/rnsupermap/JSAMQPManager.java
package com.supermap.rnsupermap;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.supermap.messagequeue.AMQPManager;
import com.supermap.messagequeue.AMQPReceiver;
import com.supermap.messagequeue.AMQPSender;

import java.util.Calendar;
import java.util.HashMap;
import java.util.Map;

/**
 * Created by Myself on 2017/5/15.
 */

public class JSAMQPManager extends ReactContextBaseJavaModule {
    public static final String REACT_CLASS = "JSAMQPManager";
    public static Map<String,AMQPManager> mManagerList=new HashMap<String,AMQPManager>();
    AMQPManager mManager;

    public JSAMQPManager(ReactApplicationContext context){super(context);}

    public static String registerId(AMQPManager manager){
        if(!mManagerList.isEmpty()) {
            for(Map.Entry entry:mManagerList.entrySet()){
                if(manager.equals(entry.getValue())){
                    return (String)entry.getKey();
                }
            }
        }

        Calendar calendar=Calendar.getInstance();
        String id=Long.toString(calendar.getTimeInMillis());
        mManagerList.put(id,manager);
        return id;
    }

    @Override
    public String getName(){
        return REACT_CLASS;
    }

    @ReactMethod
    public void createObj(Promise promise){
        try{
            AMQPManager manager = new AMQPManager();
            String _AMQPManagerId = registerId(manager);

            WritableMap map = Arguments.createMap();
            map.putString("_AMQPManagerId",_AMQPManagerId);
            promise.resolve(map);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void newReceiver(String managerId,String queueName,Promise promise){
        try{
            mManager = mManagerList.get(managerId);
            AMQPReceiver receiver = mManager.newReceiver(queueName);
            String receiverId = JSAMQPReceiver.registerId(receiver);

            WritableMap map = Arguments.createMap();
            map.putString("AMQPReceiverId",receiverId);
            promise.resolve(map);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void newSender(String managerId,Promise promise){
        try{
            mManager = mManagerList.get(managerId);
            AMQPSender sender = mManager.newSender();
            String senderId = JSAMQPSender.registerId(sender);

            WritableMap map = Arguments.createMap();
            map.putString("AMQPSenderId",senderId);
            promise.resolve(map);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void connection(String managerId, ReadableMap jsonObject, Promise promise){
        try{
            mManager = mManagerList.get(managerId);
            String ip = jsonObject.getString("IP");
            int port = jsonObject.getInt("Port");
            String hostName = jsonObject.getString("HostName");
            String userName = jsonObject.getString("UserName");
            String passWord = jsonObject.getString("PassWord");
            String clientId = jsonObject.getString("ClientId");
            boolean isConnection = mManager.connection(ip,port,hostName,userName,passWord,clientId);

            WritableMap map = Arguments.createMap();
            map.putBoolean("isConnection",isConnection);
            promise.resolve(map);
        }catch (Exception e){
            promise.reject(e);
        }
    }

    @ReactMethod
    public void disconnection(String managerId, Promise promise){
        try{
            mManager = mManagerList.get(managerId);
            mManager.disconnection();

            promise.resolve(true);
        }catch (Exception e){
            promise.reject(e);
        }
    }
}
